### ISSUE TITLE
Symlinks not created for `file:` specifiers

### ISSUE THREAD

**Issue by borekb** [lerna/lerna#1679]
When I use `file:...` specifiers for dependencies, symlinks are not created for me. For example, with the following `package.json` in `packages/app`:

```
{
  "name": "@demo/app",
  "version": "1.0.0",
  "private": true,
  "dependencies": {
    "@demo/lib": "file:../lib"
  }
}
```

running `lerna bootstrap` does _not_ create a symlink from `packages/app/node_modules/lib` to `packages/lib`.

## Expected Behavior

I'd expect a symlink to be created.

As a side note, I didn't know is that `npm` now also creates symlinks by default for local packages. So when I delete the entire `dependencies` section from the example above and run `npm i ../lib`, it will actually set up a symlink similarly to what I'd expect from Lerna. But then, when I `rm -rf packages/app/node_modules` and run `lerna bootstrap` again, the symlink is missing.

## Current Behavior

Symlink is not created.

## Steps to Reproduce

I've created a branch in my playground repo that reproduces it:

```
git clone https://github.com/borekb/lerna-playground
cd lerna-playground
git checkout -b file-specifiers
npm i
npx lerna bootstrap
```

In short, this works:

https://github.com/borekb/lerna-playground/blob/68e5b3c9f1270e028770628cce50f8caaf118231/packages/app/package.json#L6

This doesn't:

https://github.com/borekb/lerna-playground/blob/718f4a9282402ea6244d1aa7df117eb5f91d3822/packages/app/package.json#L6

## Context

We have a monorepo with internal packages that we don't publish (but we still see a lot of value in Lerna!) so the `file:` specifiers are a natural thing to use for us. Furthermore, `npm` and `yarn` would simply not work with something like `"@demo/lib": "^1.0.0"` as that package does not exist in a repository.

So currently, `bootstrap` not setting up symlinks is a pretty serious issue for us. I'm new to Lerna so maybe I'm doing some stupid mistake but I don't currently see it.

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.4.0 |
| `npm --version`  | 6.4.1|
| `node --version` | 10.10.0 |

| OS | Version |
| --- | --- |
| NAME | VERSION |
| Windows 10 | 1804 |

**Comment by evocateur** [comment]
If you are using relative `file:` specifiers, you don't need `lerna bootstrap`. Run `lerna link convert` once and you should be in business, `npm install` in the root will do everything you need (except build the packages, which you could do from a root `prepare` lifecycle with `@pectin/cli`).

**Comment by evocateur** [comment]
Quite frankly, if you're never publishing _any_ packages in your monorepo, I don't recommend using Lerna. Just use relative imports.

That being said, if you use relative file specifiers you need to also have a local file dependency in the root. You also need to run any prepare or prepublish scripts manually, as npm does not do that for local file specifiers.

TL;DR: Until you're prepared to run `lerna link convert`, don't use `file:` specifiers.

**Comment by borekb** [comment]
> That being said, if you use relative file specifiers you need to also have a local file dependency in the root.

Do you mean [like this](https://github.com/borekb/lerna-playground/blob/d91da201dc2efed5383747834323fe9da84ed5fe/package.json)? Whatever I try, it doesn't work which is strange because as you mentioned in https://github.com/lerna/lerna/issues/878#issuecomment-422096529, you use file specifiers for Lerna itself without any issues.

I think the example repo is pretty small and clear, and all `package.json`s are "normal" ([app's package.json](https://github.com/borekb/lerna-playground/blob/d91da201dc2efed5383747834323fe9da84ed5fe/packages/app/package.json), [lib's package.json](https://github.com/borekb/lerna-playground/blob/d91da201dc2efed5383747834323fe9da84ed5fe/packages/lib/package.json)) so I'm really not sure what I'm doing wrong. From your responses, I'm not sure if you think I've encountered a bug / missing behavior or not. I can see how `lerna link convert` or relative imports would avoid the issue but we can't use relative imports due to the way apps are exported from our monorepo, and the same can be said about `convert`. Is it unreasonable to expect `file:` specifiers to work in the example repo?

**Comment by RikkiGibson** [comment]
We have a monorepo which takes a dependency on a package outside the monorepo. We'd like to be able to work with a local development version of that outside package by changing the version specifier to a `file:` reference. Is that a reasonable use case? Is there an alternative approach we should be using?

**Comment by stevemao** [comment]
lerna doesn't install any `file:` deps and removes the deps in my `package-lock.json` file

**Comment by RikkiGibson** [comment]
Does the file: dep refer to a package within the monorepo @stevemao?

We use file: references to packages outside the monorepo as an alternative to `npm link`, i.e. only for development purposes. Is there any chance of this being supported? Would be interested in contributing if the maintainers are open to this feature.

**Comment by evocateur** [comment]
@borekb Yes, root `file:packages/foo` specifiers should work. All your example packages are private, it doesn't look like you _need_ Lerna?

@RikkiGibson No, lerna does not support that.

**Comment by evocateur** [comment]
@borekb oh, and there shouldn't be a lockfile in `packages/app/`, it's all handled in the root lockfile.

**Comment by RikkiGibson** [comment]
Is there a lerna-friendly workflow you'd recommend for debugging local versions of packages? In particular, using local versions of packages where it wouldn't make sense to put it in your monorepo because you don't own them or have publish rights to them?

I apologize if I'm derailing this issue by talking about this specific use case. Would be happy to open a different issue if that would be better.

**Comment by evocateur** [comment]
@RikkiGibson That's what `npm link` is for (if isolated tests in the external package aren't sufficient).

**Comment by stevemao** [comment]
@RikkiGibson no, mine is within the monorepo

**Comment by borekb** [comment]
@evocateur Thanks for the hints, will try that.

> All your example packages are private, it doesn't look like you _need_ Lerna?

Lerna does many nice things even if we never used its publish command. The two main things we wanted to use it for were:

1. Symlinking packages for local development (can probably be replaced by Yarn workspaces if one uses Yarn, which we currently are but wanted to move back to `npm`)
2. Utilize `lerna run` / `exec`.

**Comment by emmanuelbuah** [comment]
We run into the same issue. @evocateur I assume there are a number of projects out there with a mix of private and public repos. Is there a reason why symlinking for `file:...` isn't handled the same way as other package sources?

**Comment by evocateur** [comment]
Using relative `file:` specifiers, lerna is doing _nothing_ related to symlinks. That's all npm.

**Comment by wbern** [comment]
I see two alternatives for me:

1. Specify the file specifier in the root, and use `lerna bootstrap --hoist`, that has worked for me historically, despite I had no idea that this was so significant (I was just trying to get things to work at some early point).

2. Start committing all the dist/ files generated from compiled packages, and use the packages without their devDependencies.. Unless you want to do some work from a package, then you `cd` into the package, do `npm install`, and build new dist/ files that will be used automatically via the aforementioned symlinking being done beforehand. The downside of this is that it's more tedious, and committing dist files is not a huge hit here in my company either.

**Comment by wbern** [comment]
@evocateur 😢

**Comment by evocateur** [comment]
I don't know why the stale bot closed this, I feel like it's misbehaving.

@wbern `lerna bootstrap --hoist` is incompatible with relative `file:` specifiers. Use one or the other. I prefer the latter, personally.

**Comment by preco21** [comment]
I have no idea if this is correct workarounds: `lerna bootstrap && lerna exec npm install`, but it magically fixed the issue by having extra `install` command after bootstrap.

I expected that `lerna` runs `npm install` for each sub-package which basically just runs `npm` command on behalf of the tool, so `npm install` is supposed to handle `file:` dependencies as well.

* node `v12.6.0`
* npm `v6.10.2`
* lerna `v3.15.0`

**Comment by wbern** [comment]
@preco21 running npm install per package doesn't scale well. Lerna doesn't do this when you run bootstrap.

This is why I moved away from lerna to rush, because they integrate better with yarn/npm/pnpm

**Comment by evocateur** [comment]
@wbern I wish everyone would move to `rush`. It gets tiresome supporting this project all by myself for free.

**Comment by wbern** [comment]
@evocateur if lerna is crashing just be open about it.

**Comment by aalexgabi** [comment]
@evocateur I can understand that. I also understand users' frustration when they need to find where the problem comes from. I think one solution would be to manage expectations:
- clearly document known limitations in a visible part of the README
- throw an error when using `lerna bootstrap` if relative file:../module specifiers are used in package.json

In my case I spent couple of hours trying to debug what is happening as I'm migrating a project.

**Comment by wbern** [comment]
Did Lerna 3 aim to resolve this? Looking at this comment here of lerna 3:

https://github.com/lerna/lerna/issues/1312#issuecomment-371340336
